### PR TITLE
[go_router_builder]  Add go_router StatefulShellRoute support to go_router_builder

### PR DESCRIPTION
**Pull request by hangyujin**
fixes: https://github.com/flutter/flutter/issues/127371

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/go_router_builder/lib/src/route_config.dart [hunk 1]
```diff
@@ -63,6 +63,58 @@ extension $_extensionName on $_className {
...
+  String get routeDataClassName => 'StatefulShellRouteData';
+}
+
+/// The configuration to generate class declarations for a StatefulShellBranchData.
+class StatefulShellBranchConfig extends RouteBaseConfig {
+  StatefulShellBranchConfig._({
+    required this.navigatorKey,
+    required super.routeDataClass,
+    required super.parent,
+    required super.parentNavigatorKey,
```
Comment (chunhtai) [inline packages/go_router_builder/lib/src/route_config.dart:101]: StatefulShellBranch doesn't support parent navigator key

File: packages/go_router_builder/lib/src/route_config.dart [hunk 2]
```diff
@@ -63,6 +63,58 @@ extension $_extensionName on $_className {
   String get routeDataClassName => 'ShellRouteData';
 }
 
+/// The configuration to generate class declarations for a StatefulShellRouteData.
+class StatefulShellRouteConfig extends RouteBaseConfig {
+  StatefulShellRouteConfig._({
+    required this.navigatorKey,
```
Comment (chunhtai) [inline packages/go_router_builder/lib/src/route_config.dart:69]: StatefulShellRoute doesn't support navigator key

File: packages/go_router_builder/lib/src/route_config.dart [hunk 3]
```diff
@@ -63,6 +63,58 @@ extension $_extensionName on $_className {
...
+        '''
+extension $_extensionName on $_className {
+  static $_className _fromState(GoRouterState state) => const $_className();
+}
+'''
+      ];
+
+  @override
+  String get routeConstructorParameters =>
+      navigatorKey == null ? '' : 'navigatorKey: $navigatorKey,';
```
Comment (chunhtai) [inline packages/go_router_builder/lib/src/route_config.dart:89]: no navigator key, but we need to have a way to pass in
`navigatorContainerBuilder` and `restorationScopeId`

File: packages/go_router_builder/lib/src/route_config.dart [hunk 4]
```diff
@@ -356,43 +408,75 @@ abstract class RouteBaseConfig {
...
+            classElement,
+            keyName: r'$navigatorKey',
+          ),
+          parentNavigatorKey: _generateNavigatorKeyGetterCode(
+            classElement,
+            keyName: r'$parentNavigatorKey',
+          ),
+        );
+        break;
+      default:
```
Comment (chunhtai) [inline packages/go_router_builder/lib/src/route_config.dart:454]: why not use `case 'TypedGoRoute':`?
it should throw if it is anything else.

File: packages/go_router_builder/lib/src/route_config.dart [hunk 5]
```diff
@@ -356,43 +408,75 @@ abstract class RouteBaseConfig {
...
+            classElement,
+            keyName: r'$parentNavigatorKey',
+          ),
+        );
     }
 
-    value._children.addAll(reader.read('routes').listValue.map<RouteBaseConfig>(
-        (DartObject e) => RouteBaseConfig._fromAnnotation(
+    value._children.addAll(reader
+        .read(typeName == 'TypedStatefulShellRoute' ? 'branches' : 'routes')
```
Comment (chunhtai) [inline packages/go_router_builder/lib/src/route_config.dart:477]: can this be a getter in the RouteBaseConfig? In general we should try to have less check like this in this method as possible.

File: packages/go_router_builder/lib/src/route_config.dart (hunk 3 above)
Comment (hangyujin) [reply packages/go_router_builder/lib/src/route_config.dart]: done

File: packages/go_router_builder/lib/src/route_config.dart (hunk 4 above)
Comment (hangyujin) [reply packages/go_router_builder/lib/src/route_config.dart]: done

File: packages/go_router_builder/lib/src/route_config.dart (hunk 5 above)
Comment (hangyujin) [reply packages/go_router_builder/lib/src/route_config.dart]: done

File: packages/go_router_builder/lib/src/route_config.dart [hunk 6]
```diff
@@ -493,11 +577,21 @@ RouteBase get $_routeGetterName => ${_invokesRouteConstructor()};
     final String routesBit = _children.isEmpty
         ? ''
         : '''
-routes: [${_children.map((RouteBaseConfig e) => '${e._invokesRouteConstructor()},').join()}],
+${routeDataClassName == 'StatefulShellRouteData' ? 'branches' : 'routes'}: [${_children.map((RouteBaseConfig e) => '${e._invokesRouteConstructor()},').join()}],
```
[the remark this answers is not in the document]
Comment (hangyujin) [reply packages/go_router_builder/lib/src/route_config.dart]: done

File: packages/go_router_builder/lib/src/route_config.dart [hunk 7]
```diff
@@ -493,11 +577,21 @@ RouteBase get $_routeGetterName => ${_invokesRouteConstructor()};
...
 ''';
     final String parentNavigatorKeyParameter = parentNavigatorKey == null
         ? ''
         : 'parentNavigatorKey: $parentNavigatorKey,';
+
+    if (routeDataClassName == 'StatefulShellBranchData') {
```
[the remark this answers is not in the document]
Comment (hangyujin) [reply packages/go_router_builder/lib/src/route_config.dart]: done

File: packages/go_router_builder/lib/src/route_config.dart [hunk 8]
```diff
@@ -406,32 +524,36 @@ abstract class RouteBaseConfig {
   /// The parent of this route config.
   final RouteBaseConfig? parent;
 
-  /// The parent navigator key string that is used for initialize the
-  /// `RouteBase` class this config generates.
-  final String? parentNavigatorKey;
+  static String _generateChildrenGetterName(String name) {
```
[the remark this answers is not in the document]
Comment (hangyujin) [reply packages/go_router_builder/lib/src/route_config.dart:536]: it's used in the factory function `factory RouteBaseConfig._fromAnnotation`  and Instance members can’t be accessed from a factory constructor.

File: packages/go_router_builder/lib/src/route_config.dart [hunk 9]
```diff
@@ -355,44 +442,75 @@ abstract class RouteBaseConfig {
...
+        value = GoRouteConfig._(
+          path: pathValue.stringValue,
+          name: nameValue.isNull ? null : nameValue.stringValue,
+          routeDataClass: classElement,
+          parent: parent,
+          parentNavigatorKey: _generateParameterGetterCode(
+            classElement,
+            parameterName: r'$parentNavigatorKey',
+          ),
         );
```
[the remark this answers is not in the document]
Comment (hangyujin) [reply packages/go_router_builder/lib/src/route_config.dart:513]: done

File: packages/go_router_builder/lib/src/route_config.dart [hunk 10]
```diff
@@ -57,20 +59,96 @@ extension $_extensionName on $_className {
...
+
+  @override
+  String get dataConvertionFunctionName => '\$route';
+}
+
+/// The configuration to generate class declarations for a StatefulShellRouteData.
+class StatefulShellRouteConfig extends RouteBaseConfig {
+  StatefulShellRouteConfig._({
+    required super.routeDataClass,
+    required super.parent,
```
[the remark this answers is not in the document]
Comment (hangyujin) [reply packages/go_router_builder/lib/src/route_config.dart:82]: done

**Review by chunhtai — APPROVED** [review]
LGTM, nice work!

**Review by koji-1009 — COMMENTED** [review]
Sorry, I mistake this.

**Pull request merged**